packages/SwingSet/src/kernel/loadVat.js

```diff
@@ -0,0 +1,278 @@
+   *        `buildRootObject()` API, which uses LiveSlots to enforce the vat
+   *        invariants automatically.  Defaults to false.
+   *
+   *    'enablePipelining' If true, permits the kernel to pipeline messages to
+   *        promises for which the vat is the decider directly to the vat
+   *        without waiting for the promises to be resolved.  If false, such
+   *        messages will be queued inside the kernel.  Defaults to false.
+   *
+   * @param {boolean} notifyNewVat  If true, the success or failure of the
+   *    operation will be reported in a message to the admin vat, citing the
+   *    vatID
+   * @param {boolean} isDynamic  If true, the vat being created is a dynamic vat;
+   *    if false, it's a static vat (these have differences in their allowed
+   *    options and some of their option defaults).
+   *
+   * @returns {string} the vatID for a newly created vat.
+   */
+  function create(vatID, source, options, notifyNewVat, isDynamic) {
+    assert(source.bundle || source.bundleName, 'broken source');
+    const vatSourceBundle =
+      source.bundle || kernelKeeper.getBundle(source.bundleName);
+    if (!vatSourceBundle) {
+      throw Error(`Bundle ${source.bundleName} not found`);
+    }
+    // TODO: maybe hash the bundle object somehow for the description
+    const description = source.bundle
+      ? '(source bundle)'
+      : `from: ${source.bundleName}`;
+
+    assertKnownOptions(
+      options,
+      isDynamic ? allowedDynamicOptions : allowedStaticOptions,
+    );
+    const {
+      metered = isDynamic,
+      vatParameters = {},
+      enableSetup = false,
+      enablePipelining = false,
+    } = options;
+    let terminated = false;
+
+    function notifyTermination(error) {
+      if (terminated) {
+        return;
+      }
+      terminated = true;
+      const vatAdminVatId = vatNameToID('vatAdmin');
+      const vatAdminRootObjectSlot = makeVatRootObjectSlot();
+
+      const args = {
+        body: JSON.stringify([
+          vatID,
+          error
+            ? { '@qclass': 'error', name: error.name, message: error.message }
+            : { '@qclass': 'undefined' },
+        ]),
+        slots: [],
+      };
+
+      queueToExport(
+        vatAdminVatId,
+        vatAdminRootObjectSlot,
+        'vatTerminated',
+        args,
+        'logFailure',
+      );
+    }
+
+    async function build() {
+      if (typeof vatSourceBundle !== 'object') {
+        throw Error(`vat creation requires a bundle, not a plain string`);
+      }
+
+      kernelSlog.addVat(vatID, true, description);
+      const managerOptions = {
+        bundle: vatSourceBundle,
+        metered,
+        enableSetup,
+        enablePipelining,
+        enableInternalMetering: !isDynamic,
+        notifyTermination: metered ? notifyTermination : undefined,
+        vatConsole: makeVatConsole(vatID),
+        vatParameters,
+      };
+      const manager = await vatManagerFactory(vatID, managerOptions);
```

Comment:
We also lost the `kernelSlog.startup() / finish()` logging that was in `addGenesisVat` (and which wasn't present, but should have been, in the dynamic vat creation code). Please wrap the `await vatManagerFactory()` call like:

```js
      const finish = kernelSlog.startup(vatID);
      const manager = await vatManagerFactory(vatID, managerOptions);
      finish();
      addVatManager(vatID, manager, managerOptions);
```

That will capture any `console.log` calls made during top-level code (and `buildRootObject()`) in the slog, correlated with the vatID.